ratik/src/intTest/java/org/opencadc/ratik/InventoryValidatorTest.java

```diff
@@ -0,0 +1,759 @@
+        Assert.assertNotNull("DeletedStorageLocationEvent not found", localDeletedStorageLocationEvent);
+    }
+
+    /** discrepancy: artifact not in L && artifact in R
+     *
+     * explanation0: filter policy at L changed to include artifact in R
+     * evidence: ?
+     * action: equivalent to missed Artifact event (explanation3 below)
+     *
+     * explanation6: deleted from L, lost DeletedArtifactEvent
+     * evidence: ?
+     * action: assume explanation3
+     *
+     * explanation7: L==storage, deleted from L, lost DeletedStorageLocationEvent
+     * evidence: ?
+     * action: assume explanation3
+     *
+     * explanation3: L==storage, new Artifact in R, pending/missed new Artifact event in L
+     * evidence: ?
+     * action: insert Artifact
+     *
+     * L == storage
+     * before: Artifact not in L, in R
+     * after: Artifact in L
+     */
+    @Test
+    public void explanation3_ArtifactNotInLocal_LocalIsStorage() throws Exception {
+        Artifact artifact = new Artifact(URI.create("cadc:INTTEST/one.ext"), TestUtil.getRandomMD5(),
+                                         new Date(), 1024L);
+        this.remoteEnvironment.artifactDAO.put(artifact);
+
+        try {
+            System.setProperty("user.home", TMP_DIR);
+            InventoryValidator testSubject = new InventoryValidator(this.localEnvironment.daoConfig, TestUtil.LUSKAN_URI,
+                                                                    new IncludeArtifacts(),null,
+                                                                    false);
+            testSubject.run();
+        } finally {
+            System.setProperty("user.home", USER_HOME);
+        }
+
+        Artifact localArtifact = this.localEnvironment.artifactDAO.get(artifact.getID());
+        Assert.assertNotNull("local artifact not found", localArtifact);
+        Assert.assertEquals("metaChecksum mismatch", artifact.getMetaChecksum(), localArtifact.getMetaChecksum());
+    }
+
+    /** discrepancy: artifact not in L && artifact in R
+     *
+     * explanation4: L==global, new Artifact in R, pending/missed changed Artifact event in L
+     * evidence: Artifact in local db but siteLocations does not include remote siteID
+     * action: add siteID to Artifact.siteLocations
+     *
+     * L == global
+     * before: Artifact in L & R, R siteID not in L Artifact.siteLocations
+     * after:  Artifact in L and R siteID in Artifact.siteLocations
+     */
+    @Test
+    public void explanation4_ArtifactNotInL_LocalIsGlobal() throws Exception {
+        UUID remoteSiteID = this.remoteEnvironment.storageSiteDAO.list().iterator().next().getID();
+
+        Artifact artifact = new Artifact(URI.create("cadc:INTTEST/one.ext"), TestUtil.getRandomMD5(),
+                                         new Date(), 1024L);
+        this.remoteEnvironment.artifactDAO.put(artifact);
+        this.localEnvironment.artifactDAO.put(artifact);
```

Comment:
For the artifact in the localEnvironment (global), I would add a different SiteLocation. reason: in global, there should always be at least 1 SiteLocation and I'd keep the test scenario consistent with that. It's possible to enforce that in the DAO and I might add that constraint at some point before we go to live production, so that could in theory break tests that don't do  setup right.

It also enables you to verify that the code did an "add siteID" rather than "set siteID"... implying that the local artifact goes from one to two so there should be another of the last assert to check that.

Reply:
Added another SiteLocation to the local Artifact, and verify it's present after the validate has run.